---
title: "Conserved Differentially Methylated Elements from One Generation to the Next: Inheritance versus Randomness"
author: Astrid Deschenes, Pascal Belleau and Arnaud Droit 
output:
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{Permutation Positioning}
  %\VignettePackage{methylInheritance}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
library(knitr)
```

<br />
**Package**: `r Rpackage("methylInheritance")`<br />
**Authors**: `r packageDescription("methylInheritance")[["Author"]]`<br />
**Version**: `r packageDescription("methylInheritance")$Version`<br />
**Compiled date**: `r Sys.Date()`<br />
**License**: `r packageDescription("methylInheritance")[["License"]]`<br />

# Licensing 

The **methylInheritance** package and the underlying **methylInheritance** code 
are distributed under the Artistic license 2.0. You are free to use and 
redistribute this software. 

# Introduction

The analysis of treatment effects on DNA methylation, from one generation to 
the next (inter-generational) and across generations that were not exposed to 
the initial environment (trans-generational) represent complex designs. 
High-throughput sequencing techniques enable genome-wide detection of 
differentially methylated elements (DME), commonly sites (DMS) or 
regions (DMR).Due to software design, the detection of DME is usually 
made on each generation separately. However, the common DME between 
generations due to randomness is not negligible when the number of DME 
detected in each generation is high. To judge the effect on DME that is 
inherited from a treatment in previous generation, the observed number of 
conserved DME must be compared to the randomly expected number. 

We present a permutation analysis, based on Monte Carlo sampling, aim to infer 
a relation between the number of conserved DME from one generation to the next 
to the inheritance effect of treatment and to dismiss stochastic effect.

The `r Rpackage("methylInheritance")` package can perform a permutation 
analysis on both differentially methylated sites (DMS) and differentially 
methylated tiles (DMT) using the `r Biocpkg("methylKit")`.

# Loading methylInheritance package

As with any R package, the `r Rpackage("methylInheritance")` package should 
first be loaded with the following command:

```{r loadingPackage, warning=FALSE, message=FALSE}
library(methylInheritance)
```

```{r loadingFiles, warning=FALSE, message=FALSE}
F1_control_01 <- system.file("extdata",
    "F1_control_01_Cpg.txt", package = "methylInheritance")
F1_control_02 <- system.file("extdata",
    "F1_control_02_Cpg.txt", package = "methylInheritance")
F1_control_03 <- system.file("extdata",
    "F1_control_03_Cpg.txt", package = "methylInheritance")
F1_case_01 <- system.file("extdata",
    "F1_case_01_Cpg.txt", package = "methylInheritance")
F1_case_02 <- system.file("extdata",
    "F1_case_02_Cpg.txt", package = "methylInheritance")
F1_case_03 <- system.file("extdata",
    "F1_case_03_Cpg.txt", package = "methylInheritance")
F2_control_01 <- system.file("extdata",
    "F2_control_01_Cpg.txt", package = "methylInheritance")
F2_control_02 <- system.file("extdata",
    "F2_control_02_Cpg.txt", package = "methylInheritance")
F2_control_03 <- system.file("extdata",
    "F2_control_03_Cpg.txt", package = "methylInheritance")
F2_case_01 <- system.file("extdata",
    "F2_case_01_Cpg.txt", package = "methylInheritance")
F2_case_02 <- system.file("extdata",
    "F2_case_02_Cpg.txt", package = "methylInheritance")
F2_case_03 <- system.file("extdata",
    "F2_case_03_Cpg.txt", package = "methylInheritance")
F3_control_01 <- system.file("extdata",
    "F3_control_01_Cpg.txt", package = "methylInheritance")
F3_control_02 <- system.file("extdata",
    "F3_control_02_Cpg.txt", package = "methylInheritance")
F3_control_03 <- system.file("extdata",
    "F3_control_03_Cpg.txt", package = "methylInheritance")
F3_case_01 <- system.file("extdata",
    "F3_case_01_Cpg.txt", package = "methylInheritance")
F3_case_02 <- system.file("extdata",
    "F3_case_02_Cpg.txt", package = "methylInheritance")
F3_case_03 <- system.file("extdata",
    "F3_case_03_Cpg.txt", package = "methylInheritance")

F3_case_02
```

```{r}

conditions <- c(0,0,0,1,1,1)

dataNames <- c("F1A","F1B","F1C","F1D","F1E","F1F",
               "F2A","F2B","F2C","F2D","F2E","F2F",
               "F3A","F3B","F3C","F3D","F3E","F3F")

allFilesByGeneration <- list(F1=list(F1_control_01, F1_control_02, 
                                        F1_control_03, F1_case_01, 
                                        F1_case_02, F1_case_03), 
                             F2=list(F2_control_01, F2_control_02, 
                                        F2_control_03, F2_case_01, 
                                        F2_case_02, F2_case_03), 
                             F3=list(F3_control_01, F3_control_02, 
                                        F3_control_03, F3_case_01, 
                                        F3_case_02, F3_case_03))
conditionsByGeneration <- list(F1=conditions, F2=conditions, F3=conditions)
genomeVersion <- "rnor5"
nbrCores <- 1
nbrCoresDiffMeth <- 1
type <- "sites"

## Run analysis with real data (without permutation)
# runRealAnalysis(allFilesByGeneration, conditionsByGeneration,
#                                 output_dir = "real_results",
#                                 genomeVersion = genomeVersion,
#                                 nbrCores = nbrCores,
#                                 nbrCoresDiffMeth = nbrCoresDiffMeth,
#                                 doingSites = doingSites,
#                                 doingTiles = doingTiles,
#                                 minReads = 10,
#                                 minMethDiff = 20,
#                                 qvalue = 0.01,
#                                 maxPercReads = 99.9,
#                                 destrand = FALSE,
#                                 tileSize = 1000,
#                                 stepSize = 1000,
#                                 vSeed = 233333)

## Run analysis with real data (without permutation)
## Only 6 permutations are done but a much high number should be done
## for a real analysis
# a <- runPermutation(allFilesByGeneration, conditionsByGeneration,
#                                 output_dir = "test_results",
#                                 genomeVersion = genomeVersion,
#                                 nbrPermutations = 2,
#                                 nbrCores = nbrCores,
#                                 nbrCoresDiffMeth = nbrCoresDiffMeth,
#                                 doingSites = doingSites,
#                                 doingTiles = doingTiles,
#                                 minReads = 10,
#                                 minMethDiff = 20,
#                                 qvalue = 0.01,
#                                 maxPercReads = 99.9,
#                                 destrand = FALSE,
#                                 tileSize = 1000,
#                                 stepSize = 1000,
#                                 vSeed = 4444432)
# 
# 
# data <- extractData(realAnalysis_output_dir = "real_results", 
#                 permutations_output_dir = "test_results", 
#                 doingSites = doingSites, doingTiles = doingTiles)

#data
```

